storage/remote/queue_manager.go

```diff
@@ -737,26 +812,70 @@ func (s *shards) stop() {
 	}
 }
 
-// enqueue a sample.  If we are currently in the process of shutting down or resharding,
+// bufferPendingSamples listens for incoming pending samples that are caused
+// due to closing of queues. They are stored in a temporary buffer before
+// being appended to the new shards.
+func (s *shards) bufferPendingSamples() {
+	var (
+		wg             sync.WaitGroup
+		bufLock        sync.Mutex
+		buffer         []sample
+		oldShardsCount = len(*s.queueBuf)
+	)
+	if oldShardsCount == 0 {
+		return
+	}
+	wg.Add(oldShardsCount)
+	for i := 0; i < oldShardsCount; i++ {
+		go func(i int) {
+			buf := *s.queueBuf
+			temp, ok := <-buf[i]
+			if !ok {
+				wg.Done()
+				return
+			}
+			bufLock.Lock()
+			buffer = append(buffer, temp...)
+			bufLock.Unlock()
+			wg.Done()
+		}(i)
+	}
+	wg.Wait()
+	sort.SliceStable(buffer, func(i, j int) bool {
```

Comment:
I used sorting since we wanted to ensure that samples should be sorted always in the queues. But if you feel that the will eventually be sorted during streaming, then sorting can be avoided. For now, going with not using the sorts.

Reply:
We anyways need to sort, else I am not sure if sortedness is guaranteed.